cartoframes/dataset.py

```diff
@@ -55,32 +58,37 @@ def __init__(self, table_name=None, schema='public', query=None, df=None, gdf=No
             self.normalized_column_names = _normalize_column_names(gdf)
 
     @classmethod
-    def from_table(cls, table_name, schema='public', context=None):
-        return cls(
-            table_name=table_name, schema=schema, context=context, state=cls.STATE_REMOTE)
+    def from_table(cls, table_name, context, schema='public'):
+        return cls(table_name=table_name, schema=schema, context=context, state=cls.STATE_REMOTE)
 
     @classmethod
-    def from_query(cls, query, context=None):
-        return cls(
-            query=query, context=context, state=cls.STATE_REMOTE)
+    def from_query(cls, query, context):
+        return cls(query=query, context=context, state=cls.STATE_REMOTE)
 
     @classmethod
-    def from_dataframe(cls, df, table_name=None, schema='public', context=None):
-        return cls(
-            df=df, table_name=table_name, schema=schema, context=context, state=cls.STATE_LOCAL)
+    def from_dataframe(cls, df):
+        return cls(df=df, state=cls.STATE_LOCAL)
 
     @classmethod
-    def from_geodataframe(cls, gdf, table_name=None, schema='public', context=None):
-        return cls(
-            gdf=gdf, table_name=table_name, schema=schema, context=context, state=cls.STATE_LOCAL)
+    def from_geodataframe(cls, gdf):
+        return cls(gdf=gdf, state=cls.STATE_LOCAL)
 
     @classmethod
-    def from_geojson(cls, geojson, table_name=None, schema='public', context=None):
-        return cls(
-            gdf=load_geojson(geojson), table_name=table_name, schema=schema, context=context, state=cls.STATE_LOCAL)
+    def from_geojson(cls, geojson):
+        return cls(gdf=load_geojson(geojson), state=cls.STATE_LOCAL)
+
+    def upload(self, with_lonlat=None, if_exists='fail', table_name=None, schema=None, context=None):
+        if table_name:
+            self.table_name = table_name
+        if schema:
+            self.schema = schema
+        if context:
+            self.cc = context
+
+        if self.table_name is None or self.cc is None:
+            raise CartoException('You should provide a table_name and context to upload data')
 
-    def upload(self, with_lonlat=None, if_exists='fail'):
-        if self.query and self.table_name is not None and not self.exists():
+        if self.query and not self.exists():
```

Comment:
What happens if I want to create a table from a query and the table already exists? And if I want to append data?

I suggest to add some test cases

Reply:
I didn't want to spend more time on it, but probably you are right because it smells like a dead rat

Reply:
I don't want to handle the objectives from https://github.com/CartoDB/cartoframes/issues/670 but we are here in this method and probably in more ones (like `download`).

In this case, we need to prioritize some properties. My initial idea is the following:
- prioritize local to remote
- between df & gdf, I think we should prioritize gdf

So the order would be:
- gdf
- df
- query

How do you see it?

Reply:
I have applied the approach described before, but I still want to hear your opinion

Reply:
And following a similar idea, I have refoctored the Dataset download method https://github.com/CartoDB/cartoframes/pull/676

I have used another PR to avoid blocks in this one

Reply:
Looks good to me tackling this in another PR. I just wanted to point it out 😅 

About the order, I'd say:

- df
- gdf
- query

I'm not sure GeoDataframe is more used than a regular pandas dataframe.

I'd say in many cases, we'll receive a pandas dataframe with the lon lat columns.

Reply:
I put gdf first because I can imagine 2 cases:
- The user only add one or another (so the order is not really important)
- The user adds a DataFrame and then creates a Map with it. In this process, we will need to create the GeoDataFrame and then the geoJSON (maybe this is not the best way)